boot/LoadReactOS.c

```diff
@@ -0,0 +1,473 @@
+			break;
+		}
+	}
+
+	if ((u8 *)mbHeader - buffer >= MULTIBOOT_SEARCHAREA_LEN) {
+		printk("No multiboot header found\n");
+		return NULL;
+	}
+
+	return mbHeader;
+}
+
+CONFIGENTRY *DetectReactOSNative(char *szGrub) {
+	CONFIGENTRY *config;
+	int nRet;
+
+	strcpy(&szGrub[4], "/freeldr.sys");
+	nRet = grub_open(szGrub);
+
+	if (nRet != 1 || errnum) {
+		// File not found
+		errnum = 0;
+		return NULL;
+	}
+
+	config = (CONFIGENTRY *)malloc(sizeof(CONFIGENTRY));
+	memset(config, 0x00, sizeof(CONFIGENTRY));
+	config->bootSystem = SYS_REACTOS;
+	strcpy(config->title, "ReactOS");
+	strcpy(config->szPath, "/freeldr.sys");
+
+	grub_close();
+	return config;
+}
+
+int LoadReactOSNative(char *szGrub, CONFIGENTRY *config) {
+	int nRet;
+	long dwSize;
+	OPTMULTIBOOT *multiboot = &config->opt.Multiboot;
+
+	VIDEO_ATTR = 0xffd8d8d8;
+	printk("  Loading %s ", config->szPath);
+	VIDEO_ATTR = 0xffa8a8a8;
+
+	strncpy(&szGrub[4], config->szPath, strlen(config->szPath));
+
+	nRet = grub_open(szGrub);
+
+	if (nRet != 1) {
+		printk("Unable to load file, Grub error %d\n", errnum);
+		errnum = 0;
+		wait_ms(2000);
+		return false;
+	}
+
+	dwSize = grub_read(FREELDR_LOAD_AREA, filemax);
+	grub_close();
+	printk(" - %d bytes\n", dwSize);
+
+	if (CheckMultibootHeader(FREELDR_LOAD_AREA) == NULL) {
+		wait_ms(2000);
+		return false;
+	}
+	multiboot->pBuffer = FREELDR_LOAD_AREA;
+	multiboot->uBufferSize = dwSize;
+	/* Multiboot partition indices starting at 0 */
+	multiboot->uBootDevice = GetMultibootBootDevice(0x80 + config->drive, config->partition, 0xFF, 0xFF);
+
+	return true;
+}
+
+CONFIGENTRY *DetectReactOSFATX(FATXPartition *partition) {
+	FATXFILEINFO fileinfo;
+	CONFIGENTRY *config = NULL;
+
+	if (LoadFATXFile(partition, "/freeldr.sys", &fileinfo)) {
+		// Root of E has a freeldr.sys in
+		config = (CONFIGENTRY *)malloc(sizeof(CONFIGENTRY));
+		memset(config, 0x00, sizeof(CONFIGENTRY));
+		config->bootSystem = SYS_REACTOS;
+		strcpy(config->title, "ReactOS");
+		strcpy(config->szPath, "/freeldr.sys");
+
+		free(fileinfo.buffer);
+	}
+
+	return config;
+}
+
+int LoadReactOSFATX(FATXPartition *partition, CONFIGENTRY *config) {
+	static FATXFILEINFO fileinfo;
+	OPTMULTIBOOT *multiboot = &config->opt.Multiboot;
+
+	memset(&fileinfo, 0x00, sizeof(fileinfo));
+
+	VIDEO_ATTR = 0xffd8d8d8;
+	printk("  Loading %s from FATX", config->szPath);
+	if (!LoadFATXFilefixed(partition, config->szPath, &fileinfo, FREELDR_LOAD_AREA)) {
+		printk("Error loading %s\n", config->szPath);
+		wait_ms(2000);
+		return false;
+	} else {
+		printk(" - %d bytes\n", fileinfo.fileRead);
+
+		if (CheckMultibootHeader(fileinfo.buffer) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = fileinfo.buffer;
+		multiboot->uBufferSize = fileinfo.fileRead;
+		/* Multiboot partition indices starting at 0, FATX (E:) is first one */
+		multiboot->uBootDevice = GetMultibootBootDevice(0x80 + partition->nDriveIndex, 0, 0xFF, 0xFF);
```

Comment:
```c
// LoadReactOSNative
multiboot->uBootDevice = GetMultibootBootDevice(0x80 + config->drive         , config->partition, 0xFF, 0xFF);
// LoadReactOSFATX
multiboot->uBootDevice = GetMultibootBootDevice(0x80 + partition->nDriveIndex, 0                , 0xFF, 0xFF);
```

How would these 2 be distinguished by freeloader (or cromwell)?

If Native `config->partition` is 0 and `config->drive` is X, wouldn't that be equivalent to FATX where `partition->nDriveIndex` is X? Can this happen? If reactOS will boot *any* filesystem (by auto-detect) then we wouldn't need 2 functions in Cromwell probably (as it's misleading to explicitly pick FATX boot, but to end up with FAT32).

Reply:
Both Cromwell and FreeLoader supports only FATX data partition (E:), that's why the value is equal to 0 in `LoadReactOSFATX()`.

FreeLoader can autodetect partition scheme and the filesystem, this information is not passed via multiboot. I checked it boots correctly, either from MBR/FAT32 or Xbox(BRFR)/FATX.

Also FreeLoader will always boot from MBR HDD when possible, and only if MBR is not detected, it will try to probe Xbox partitions. So the collision here is not possible.

Reply:
I feel like your explanation contradicts the code-comment which implies that "E:" is the first one, but then other partitions follow.

The function names also imply that this picks and depends on a partition scheme, whereas your reply suggests that this is merely an independent pre-test?

Reply:
Improved the comment to make things clear.